### PR TITLE
final v28 foundation spec

### PR DESCRIPTION
**Pull request by jannotti**
Changes to readmes and specs to reflect the matching foundation spec.

### CODE REVIEW

File: data/transactions/logic/README.md [hunk 1]
```diff
@@ -81,7 +80,7 @@ An application transaction must indicate the action to be taken following the ex
...
+Many instructions accept values to designate Accounts, Assets, or Applications. Beginning with TEAL v4, these values may always be given as an _offset_ in the corresponding Txn fields (Txn.Accounts, Txn.ForeignAssets, Txn.ForeignApps) _or_ as the value itself (a bytes address for Accounts, or a uint64 ID). The values, however, must still be present in the Txn fields. Before TEAL v4, most opcodes required the use of an offset, except for reading account local values of assets or applications, which accepted the IDs directly and did not require the ID to be present in they corresponding _Foreign_ array. (Note that beginning with TEAL v4, those ID are required to be present in their corresponding _Foreign_ array.) See individual opcodes for details. In the case of account offsets or application offsets, 0 is specially defined to Txn.Sender or the ID of the current application, respectively.
```
Comment (jdtzmn) [inline data/transactions/logic/README.md:83]: nit: (Note that beginning with TEAL v4, those ***IDs*** are [...])

File: data/transactions/logic/README.md [hunk 2]
```diff
@@ -95,7 +94,9 @@ A contract account governed by a buggy program might not have a way to get asset
 
 For one-argument ops, `X` is the last element on the stack, which is typically replaced by a new value.
 
-For two-argument ops, `A` is the previous element on the stack and `B` is the last element on the stack. These typically result in popping A and B from the stack and pushing the result.
+For two-argument ops, `A` is the penultimate element on the stack and `B` is the top of the stack. These typically result in popping A and B from the stack and pushing the result.
+
+For three-argument ops, `A` is the element two below the top, `B` is the penultimate stack element and `C` is the top of the stack. These operatiosn typically pop A, B, and C from the stack and push the result.
```
Comment (jdtzmn) [inline data/transactions/logic/README.md:99]: nit: operations

File: data/transactions/logic/README.md [hunk 3]
```diff
@@ -115,8 +116,8 @@ For two-argument ops, `A` is the previous element on the stack and `B` is the la
 | `\|\|` | A is not zero or B is not zero => {0 or 1} |
 | `shl` | A times 2^B, modulo 2^64 |
 | `shr` | A divided by 2^B |
-| `sqrt` | The largest integer X such that X^2 <= A |
-| `bitlen` | The index of the highest bit in A. If A is a byte-array, it is interpreted as a big-endian unsigned integer |
+| `sqrt` | The largest integer B such that B^2 <= X |
+| `bitlen` | The highest set bit in X. If X is a byte-array, it is interpreted as a big-endian unsigned integer. bitlen of 0 is 0, bitlen of 8 is 4 |
```
Comment (jdtzmn) [inline data/transactions/logic/README.md:120]: I'm confused by the change from A/B to X here. Isn't it convention to use A, B, etc. for stack values?

File: data/transactions/logic/doc.go [hunk 4]
```diff
@@ -198,18 +198,18 @@ var opDocExtras = map[string]string{
...
-	"gload":             "The `gload` opcode can only access scratch spaces of previous app calls contained in the current group.",
-	"gloads":            "The `gloads` opcode can only access scratch spaces of previous app calls contained in the current group.",
-	"gaid":              "The `gaid` opcode can only access the ID of assets or applications created by previous txns in the current group.",
-	"gaids":             "The `gaids` opcode can only access the ID of assets or applications created by previous txns in the current group.",
+	"gload":             "`gload` fails unless the requested transaction is an ApplicationCall and T < GroupIndex.",
+	"gloads":            "`gloads` fails unless the requested transaction is an ApplicationCall and X < GroupIndex.",
+	"gaid":              "`gaid` fails unless the requested transaction created an asset or application and T < GroupIndex.",
+	"gaids":             "`gaidx` fails unless the requested transaction created an asset or application and X < GroupIndex.",
```
Comment (jdtzmn) [inline data/transactions/logic/doc.go:204]: typo: gaids

**Review by jdtzmn — CHANGES_REQUESTED** [review]

File: data/transactions/logic/README.md (hunk 3 above)
Comment (jannotti) [reply data/transactions/logic/README.md:120]: I thought so too, but it was pointed out to me that the convention that's actually explained in the docs is that single argument opcodes use X.

Comment (jdtzmn) [reply data/transactions/logic/README.md:120]: Huh - good to know

**Review by jdtzmn — APPROVED** [review]

File: data/transactions/logic/TEAL_opcodes.md [hunk 5]
```diff
@@ -554,33 +554,33 @@ The `gloads` opcode can only access scratch spaces of previous app calls contain
...
 - Opcode: 0x3d
 - Pops: *... stack*, uint64
 - Pushes: uint64
-- push the ID of the asset or application created in the Ath transaction of the current group
+- push the ID of the asset or application created in the Xth transaction of the current group
 - LogicSigVersion >= 4
 - Mode: Application
 
-The `gaids` opcode can only access the ID of assets or applications created by previous txns in the current group.
+`gaidx` fails unless the requested transaction created an asset or application and X < GroupIndex.
```
Comment (nullun) [inline data/transactions/logic/TEAL_opcodes.md:568]: Should this be `gaids`?

**Pull request merged**